File: ADA/Lab Programs/MergeSort.cpp
#include <iostream>
using namespace std;

void merge(int arr[], int s, int mid, int e) {
    cout << "Merge(" << s << ", " << mid << ", " << e << ")\n";
    int l = mid - s + 1;
    int r = e - mid;

    int L[l];
    int R[r];

    for(int i = 0; i < l; i++)
        L[i] = arr[s + i];
    for(int j = 0; j < r; j++)
        R[j] = arr[mid + j + 1];

    // merge part
    int i = 0, j = 0, k = s;
    while(i < l && j < r) {
        if(L[i] <= R[j])
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    while(i < l)
        arr[k++] = L[i++];
    
    while(j < r)
        arr[k++] = R[j++];
}

void mergeSort(int arr[], int s, int e) {
    if(s < e) {
        cout << "MergeSort(" << s << ", " << e << ")\n";
        int mid = (s + e) / 2;

        mergeSort(arr, s, mid);
        mergeSort(arr, mid + 1, e);
        merge(arr, s, mid, e);
    }
}

int main() {
    int nums[] = {90, 2, 1, 45, 23, 5};
    mergeSort(nums, 0, 5);
    for(int i = 0; i < 6; i++)
        cout << nums[i] << " ";
    cout << "\n";

    return 0;
}

File: Python/sorting/MergeSort.py
def merge(arr, low, mid, high):
    m = mid - low + 1
    n = high - mid
    left = [0] * m
    right = [0] * n

    for i in range(0, m):
        left[i] = arr[low + i]
    for j in range(0, n):
        right[j] = arr[mid + j + 1]

    i, j, k = 0, 0, low

    while i < m and j < n :
        if left[i] <= right[j] :
            arr[k] = left[i]
            i += 1
        else :
            arr[k] = right[j]
            j += 1
        k += 1
    
    while i < m :
        arr[k] = left[i]
        i += 1
        k += 1
    while j < n :
        arr[k] = right[j]
        j += 1
        k += 1

def mergeSort(arr, low, high):
    if low < high :
        mid = (low + high) // 2
        mergeSort(arr, low, mid)
        mergeSort(arr, mid + 1, high)
        merge(arr, low, mid, high)

arr = [12, 43, 21, 53, 85]
mergeSort(arr, 0, len(arr) - 1)

for i in range(0, len(arr)):
    print(arr[i])

